ark/jsonschema/composition.ts

```diff
@@ -0,0 +1,95 @@
+import type { array } from "@ark/util"
+import { type, type Type } from "arktype"
+import { innerParseJsonSchema, type inferJsonSchema } from "./json.ts"
+import type { JsonSchema } from "./scope.ts"
+
+const validateAllOfJsonSchemas = (
+	jsonSchemas: JsonSchema.Schema[]
+): Type<unknown> =>
+	jsonSchemas
+		.map(jsonSchema => innerParseJsonSchema.assert(jsonSchema))
+		.reduce((acc, validator) => acc.and(validator))
+
+const validateAnyOfJsonSchemas = (
+	jsonSchemas: JsonSchema.Schema[]
+): Type<unknown> =>
+	jsonSchemas
+		.map(jsonSchema => innerParseJsonSchema.assert(jsonSchema))
+		.reduce((acc, validator) => acc.or(validator))
+
+const validateNotJsonSchema = (jsonSchema: JsonSchema.Schema) => {
+	const inner = innerParseJsonSchema.assert(jsonSchema)
+	return type("unknown").narrow((data, ctx) =>
+		inner.allows(data) ? ctx.mustBe(`not ${inner.description}`) : true
+	) as Type<unknown>
+}
+
+const validateOneOfJsonSchemas = (jsonSchemas: JsonSchema.Schema[]) => {
+	const oneOfValidators = jsonSchemas.map(nestedSchema =>
+		innerParseJsonSchema.assert(nestedSchema)
+	)
+	const oneOfValidatorsDescriptions = oneOfValidators.map(
+		validator => `○ ${validator.description}`
+	)
+	return (
+		(
+			type("unknown").narrow((data, ctx) => {
+				let matchedValidator: Type | undefined = undefined
+
+				for (const validator of oneOfValidators) {
+					if (validator.allows(data)) {
+						if (matchedValidator === undefined) {
+							matchedValidator = validator
+							continue
+						}
+						return ctx.mustBe(
+							`exactly one of:\n${oneOfValidatorsDescriptions.join("\n")}`
+						)
+					}
+				}
+				return matchedValidator !== undefined
+			}) as Type<unknown>
+		)
+			// TODO: Theoretically this shouldn't be necessary due to above `ctx.mustBe` in narrow???
```

Comment:
Description can be useful for introspectability, but yeah generally you'd either want to use that or a custom error in narrow if you need additional context.

Since it looks like `oneOfValidators` is already known before the type is created, it's important that stuff like `mustBe` that would run on every failed validation is just evaluated when the type is created, and shouldn't duplicate the logic of the description.